### ISSUE TITLE
Add compute shader for OpenGL and Vulkan renderer

### ISSUE THREAD

**Issue by ghsoares** [godotengine/godot-proposals#1027]
**Describe the project you are working on:**
Test project where I create a dynamic 3D water system with arrays on surface

**Describe the problem or limitation you are having in your project:**
In my project, I'm creating a dynamic water system where I use a 2D array as the surface height,
but manipulating this array from GDScript can be very slow as I increase surface quality (n^2 time complexity). This could be faster if I could manipulate variables directly in GPU (as an arbitrary choosen example, the TensorFlow library uses GPU Matrices calculation to perform the neural network tasks faster).

**Describe the feature / enhancement and how it helps to overcome the problem or limitation:**
The feature would be a new shader_type, where it would have the desired uniforms on top, and the logic on body. The uniforms would be only for initializing the logic, and the logic inside would manipulate the uniforms every frame. With this, the heavy tasks as large 2D arrays would be leaved only for GPU, as it performs more tasks that CPU.

**Describe how your proposal will work, with code, pseudocode, mockups, and/or diagrams:**
The new shader_type would be something like this:

```glsl
shader_type gpu_task; // or other name

uniform float[][] water_heights; // the current shader api dont works with arrays as uniform

// updates every frame (before rendering)
void update() {
    // The logic goes here
}

// reset variables?
void reset() {}
```

As it would perform tasks directly on GPU, it can apply those variables on a second-pass shader, where would be rendered after. So the second-pass shader would have the same variable names as the new shader.

The reason why it would be a shader_type because it would be simpler to program for those that already knows the Godot shader language.

Maybe I'm saying something impossible, but it would be nice to have in Godot.

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
I don't think so, as the current shader API's uniforms dont works with arrays and dont write then.

**Is there a reason why this should be core and not an add-on in the asset library?:**
This would increase the performance in a majority of in-game tasks, as the water update example.

**Comment by SIsilicon** [comment]
What you are describing is something like a computer shader, which is planned to be part of Godot 4.0. :)

**Comment by ghsoares** [comment]
@SIsilicon ah yes, I didn't know that compute shaders is already a milestone for Godot 4.0 . As for compatibility only for Vulkan renderer, it would be nice to have for OpenGL renderer too.

**Comment by clayjohn** [comment]
Adding compute shader support to OpenGL was discussed in #877. 

Compute shaders are already mostly supported in 4.0, more work needs to be done to make them easy to use. 

As for OpenGL, the OpenGL backend will not support a new enough OpenGL version to support Compute. Either it will support OpenGL 2.1 or OpenGL 3.3 to maximize compatibility. Neither of which supports Compute..

**Comment by SIsilicon** [comment]
The alternative would be to support other compute APIs like OpenCL and CUDA.

**Comment by clayjohn** [comment]
@SIsilicon having alternative compute APIs is a whole other can of worms. It can be useful for physics and other game systems when done right. But it can easily lead to horrible performance bottlenecks especially if you use it for rendering tasks. 

The slowest part of GPU compute these days is transferring data between CPU and GPU. Rendering gets around this by doing as much as possible in Compute and avoiding transferring the data back to the CPU at all. If we expose a GPU compute library like OpenCL or CUDA, then we get stuck having to transfer the data 3 times (1. to the GPU for compute, 2. back to the CPU, 3. to the GPU for rendering). For rendering tasks this is prohibitively slow. For physics or GPGPU, it is only 2 transfers, but you still need to be extremely careful about how it is used.

**Comment by SIsilicon** [comment]
@clayjohn I totally agree with you. Using compute API along with the rendering pipeline can be a tedious task if done incorrectly. However, it is possible to couple the two. For instance, it is possible to directly transfer data from an OpenCL context to an OpenGL one. And we already know that Vulkan's compute shader is connected with the rendering stuff anyway.

**Comment by Marenz** [comment]
I believe this issue works at the very least now with godot 4.4 and can be closed?

**Comment by Calinou** [comment]
> I believe this issue works at the very least now with godot 4.4 and can be closed?

Indeed, this is now available in Forward+ and Mobile since 4.0. See [documentation](https://docs.godotengine.org/en/latest/tutorials/shaders/compute_shaders.html) and [demo projects](https://github.com/godotengine/godot-demo-projects/tree/master/compute) ([one more](https://github.com/godotengine/godot-demo-projects/tree/master/misc/compute_shader_heightmap)).

There are no plans to implement this in Compatibility given OpenGL ES limitations (and WebGL even more so).
